File: rockpaperscissors-web/src/types/game.ts
/**
 * The game object
 */
export interface Game {
  id: string;
  roundsNumber: number;
  players: Record<string, Player>;
  rounds: Array<Round>;
  status: string;
  currentRound: number;
  winner: string;
  isDraw: boolean;
  pausedBy: string;
}

/**
 * The player object
 */
export interface Player {
  nickname: string;
}

/**
 * Round information
 */
export interface Round {
  /**
   * Map of nickname - move so we know what the players choose
   */
  playerSelections: Record<string, Move> | null;
  /**
   * The player that won the round
   */
  winner: string;
  isDraw: boolean;
}

/**
 * Selected move from the players
 */
export type Move = "ROCK" | "PAPER" | "SCISSORS";

export const Moves: Record<string, Move> = {
  rock: "ROCK",
  paper: "PAPER",
  scissors: "SCISSORS",
};

File: rockpaperscissors-api/internal/domain/game.go
package domain

import (
	"github.com/google/uuid"
	gwebsocket "github.com/gorilla/websocket"
)

// Game object
type Game struct {
	ID           string             `json:"id"`
	RoundsNumber int                `json:"roundsNumber"`
	Players      map[string]*Player `json:"players"`
	Winner       string             `json:"winner"`
	Rounds       []Round            `json:"rounds"`
	Status       GameStatus         `json:"status"`
	CurrentRound int                `json:"currentRound"`
	IsDraw       bool               `json:"isDraw"`
	PausedBy     string             `json:"pausedBy"`
}

// NewGame creates a new game with the given rounds and a generated ID
func NewGame(roundsNumber int) Game {
	if roundsNumber < 1 {
		roundsNumber = 1
	}
	return Game{
		ID:           uuid.NewString(),
		RoundsNumber: roundsNumber,
		Rounds:       make([]Round, roundsNumber),
		Players:      make(map[string]*Player),
		CurrentRound: 0,
	}
}

// Player object
type Player struct {
	Nickname string           `json:"nickname"`
	Conn     *gwebsocket.Conn `json:"-"`
}

// Round players selections and winner for a single round
type Round struct {
	PlayerSelections map[string]Move `json:"playerSelections"`
	Winner           string          `json:"winner"`
	IsDraw           bool            `json:"isDraw"`
}

// Selection move selection by a single player
type Selection struct {
	MadeBy Player
	Move   Move
}
